### PR TITLE
chore(devDeps): bump metro and react-native

### PR DESCRIPTION
**Pull request by legobeat**
## **Description**

Update react-native and metro to receive patches.
- Update metro from `0.73` to `0.76` 
- Migrate from `metro-config` to `@react-native/metro-config`
- Update `react-native` from `0.71.15` to `0.72.14`
  - Port patch
  - Remove hermes patch
- Bump and unpin `@react-native-community/` packages
- Update `@react-native-community/netinfo` from `6.0.0` to `^9.5.0`
  - [Changelog](https://github.com/react-native-netinfo/react-native-netinfo/blob/master/CHANGELOG.md)
  - [Release notes](https://github.com/react-native-netinfo/react-native-netinfo/releases)
  - v10 breaking change:
    > netinfo now requires AbortController
- Update `@react-native-masked-view/masked-view`
  - [Release notes](https://github.com/react-native-masked-view/masked-view/releases) 
- Upgrade to Gradle version 8.0.1 and Android SDK 33

React Native [Upgrade](https://github.com/orgs/MetaMask/projects/60/views/6?pane=issue&itemId=41033345)
[Android SDK Update](https://github.com/react-native-community/cli/issues/1856)

## **Related issues**

- Fixes: https://github.com/MetaMask/mobile-planning/issues/1262
- #7759

## **Manual testing steps**

Node environment must be `18.18.2`
Supports Java version `17.0.10`

1. Check out branch
2. Run `yarn setup`
3. Run `yarn watch`
4. Run `yarn start:ios` - wait for app to build on iOS simulator
5. Run `yarn start:android` - wait for app to build Android emulator

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: patches/react-native+0.72.12.patch [hunk 1]
```diff
@@ -15,7 +15,8 @@ index 1379ffd..f2825f5 100644
  const start = Date.now();
  
  require('./setUpGlobals');
-+// require('./setUpDOM'); Introduced in RN v0.72, ensure included when upgrading patch
+-require('./setUpDOM');
++// require('./setUpDOM');
```
[the remark this answers is not in the document]
Comment (leotm) [reply patches/react-native+0.72.12.patch]: good spot ^ yep can confirm it's present [upstream](https://github.com/facebook/react-native/blob/v0.72.12/packages/react-native/Libraries/Core/InitializeCore.js#L30) so we want it uncommented

[the remark this answers is not in the document]
Comment (leotm) [reply patches/react-native+0.72.12.patch]: updated our patch here https://github.com/MetaMask/metamask-mobile/pull/9280/commits/4b499b629d174e3e5aa45bbcce72257a562dfcbf

**Review by Gedoxa — APPROVED** [review]

File: package.json [hunk 2]
```diff
@@ -252,7 +252,7 @@
     "query-string": "^6.12.1",
     "randomfill": "^1.0.4",
     "react": "18.2.0",
-    "react-native": "0.71.15",
+    "react-native": "0.72.12",
```
Comment (leotm) [inline package.json:255]: spotted RN [v0.72.13](https://github.com/facebook/react-native/releases/tag/v0.72.13) release we may want to bump to while at it

Comment (leotm) [reply package.json:255]: we'll want to make these changes too https://react-native-community.github.io/upgrade-helper/?from=0.71.15&to=0.72.13&package=io.metamask&name=MetaMask cc @MetaMask/mobile-platform

File: patches/react-native+0.72.12.patch
Comment (leotm) [inline patches/react-native+0.72.12.patch:1]: tryna figure why getting these changes line 131 onwards (not picked up in https://github.com/MetaMask/metamask-mobile/pull/9280/commits/e1d12c8b3f562c636f44c5268b39ef159fabc438 https://github.com/MetaMask/metamask-mobile/pull/9280/commits/92fb6f983d9ee15cdee8d5a34a2078d482c7538f)
or if it's something that happens on Macs
or whether it's just patch-package _doing it's thing_

File: patches/react-native+0.72.12.patch [hunk 3]
```diff
@@ -122,43 +122,16 @@ index d03f3dd..816f559 100644
...
-@@ -59,7 +59,7 @@ def skipPrefabPublishing = System.getenv("REACT_NATIVE_HERMES_SKIP_PREFAB") != n
- def jsiDir = new File(reactNativeRootDir, "ReactCommon/jsi")
- 
- task downloadHermes(type: Download) {
--    src("https://github.com/facebook/hermes/tarball/${hermesVersion}")
-+    src("https://github.com/MetaMask/hermes/tarball/${hermesVersion}")
-     onlyIfNewer(true)
-     overwrite(false)
-     dest(new File(downloadsDir, "hermes.tar.gz"))
-@@ -120,12 +120,9 @@ android {
```
Comment (leotm) [inline patches/react-native+0.72.12.patch:161]: just noting here removing our `node_modules/react-native/ReactAndroid/hermes-engine/build.gradle` patch to match upstream, should be reverted along with https://reactnative.dev/contributing/how-to-build-from-source#update-your-project-to-build-from-source to switch back to bundled Hermes cc @Cal-L @jpcloureiro

(original convo [here](https://github.com/MetaMask/metamask-mobile/pull/6220#discussion_r1218890123) and more context [here](https://github.com/facebook/hermes/issues/1241))

then if we need to build from source again, current simpler way is with [`REACT_NATIVE_OVERRIDE_HERMES_DIR`](https://github.com/facebook/hermes/blob/main/doc/ReactNativeIntegration.md)

**Review by leotm — COMMENTED** [review]
i did a brief check of [reactnative.dev/docs/0.71/environment-setup](https://reactnative.dev/docs/0.71/environment-setup) vs [reactnative.dev/docs/0.72/environment-setup](https://reactnative.dev/docs/0.72/environment-setup)

couldn't spot anything we need to update in our [docs](https://github.com/MetaMask/metamask-mobile/blob/main/docs/readme/environment.md), but may have missed something cc @MetaMask/mobile-platform

nb: JDK 17 is required in RN v0.73+
nb: Android 14 (UpsideDownCake) SDK required in RN v0.73+

nb: Node 18+ required in RN v0.73+ (we're ahead on Node 20)

nb: babel.config.js: [metro-react-native-babel-preset](https://github.com/facebook/metro/tree/v0.77.0/packages/metro-react-native-babel-preset) -> [@react-native/babel-preset](https://github.com/facebook/react-native/tree/main/packages/react-native-babel-preset) in RN v0.73+

nb: RN new experimental [debugger](https://reactnative.dev/docs/0.73/debugging?js-debugger=new-debugger) preview RN v0.73+ to replace Flipper

File: package.json [hunk 4]
```diff
@@ -443,9 +442,9 @@
     "koa": "^2.14.2",
     "lint-staged": "10.5.4",
     "listr2": "^8.0.2",
-    "metro-config": "~0.73.10",
-    "metro-react-native-babel-preset": "~0.73.10",
-    "metro-react-native-babel-transformer": "~0.73.10",
+    "metro-config": "~0.76.9",
+    "metro-react-native-babel-preset": "~0.76.9",
+    "metro-react-native-babel-transformer": "~0.76.9",
```
Comment (leotm) [inline package.json:447]: we'll also want to update 

https://github.com/MetaMask/metamask-mobile/blob/5dc336c10b5c27f8ebddc3f059489620d6dc4aad/metro.config.js#L9-L14

to extend [`@react-native/metro-config`](https://github.com/facebook/react-native/blob/v0.72.13/packages/react-native/template/metro.config.js) instead

fresh `npx react-native@0.72.13 init RN07213 --version 0.72.13` warns

```console
warn From React Native 0.72, your metro.config.js file should extend'@react-native/metro-config'. Please see the React Native 0.72 changelog, or copy the template at:
https://github.com/facebook/react-native/blob/main/packages/react-native/template/metro.config.js
warn Falling back to internal defaults.
```

Comment (leotm) [reply package.json:447]: https://react-native-community.github.io/upgrade-helper/?from=0.71.15&to=0.72.13&package=io.metamask&name=MetaMask#RnDiffApp-metro.config.js

Comment (leotm) [reply package.json]: great Qu ^ it should be [rn-diff-purge](https://github.com/react-native-community/rn-diff-purge)'ing
`npx react-native@0.71.15 init RnDiffApp --version 0.71.15`
vs
`npx react-native@0.72.13 init RnDiffApp --version 0.72.13`
according to [react-native-community/upgrade-helper](https://github.com/react-native-community/upgrade-helper)

Comment (leotm) [reply package.json]: (but it's named `HelloWorld` [upstream](https://github.com/facebook/react-native/blob/v0.72.13/packages/react-native/template/app.json) for some reason)

Comment (leotm) [reply package.json]: ah yes knew it rings a bell ^ (noting it in our [TS 4.8.4 upg](https://github.com/MetaMask/metamask-mobile/pull/6080))

https://github.com/MetaMask/metamask-mobile/blob/444c0dfdad94553cd2575ea87b4bd7f7f81caf75/.eslintrc.js#L2-L9

> The one thing I can still see diffing in the upgrade helper link (whee is that manifest sourced from btw?) there is migrating from `@react-native-community/eslint-config` to `@react-native/eslint-config`. Trying that seems to break storybook, however, so I guess that would be postponed to an upcoming storybook upgrade.

imo postponing this eslint upg/migration till after this RN upg is merged should be ok and help keep both tasks smaller @brianacnguyen

**Comment by leotm** [comment]
just noting we've removed old https://github.com/facebook/react-native/blob/v0.71.15/template/_node-version
in RN 72 upstream https://github.com/facebook/react-native/tree/v0.72.14/packages/react-native/template
so we may want to remove ours too, currently here
https://github.com/MetaMask/metamask-mobile/blob/main/.nvmrc

**Comment by legobeat** [comment]
> just noting we've removed old [.node-version](https://github.com/facebook/react-native/blob/v0.71.15/template/_node-version) in RN 72 [upstream](https://github.com/facebook/react-native/tree/v0.72.14/packages/react-native/template) so we may want to remove our [.nvmrc](https://github.com/MetaMask/metamask-mobile/blob/main/.nvmrc) too to simplify our setup if not causing any issues

It's still used in GitHub Actions CI.
But in any case, it would be nice to consolidate the declaration of the project nodejs version.

We could already tell `asdf` to read version from `.nvmrc` as well: https://github.com/asdf-vm/asdf-nodejs?tab=readme-ov-file#nvmrc-and-node-version-support

They do refer to it as legacy, though. An alternative would be to replace `.nvmrc` with the newer `.tool-versions`, which `@actions/setup-node` [also supports](https://github.com/actions/setup-node/blob/c2ac33f2c62f978d6c944d9648125a294e56dc0b/action.yml#L11).

**Comment by Daniel-Cross** [comment]
I'm looking into this again now the Hackathon is over. Managed to get the branch in sync with `main` and build on iOS but I get the following issue with @walletconnect:
Simulator Screenshot - iPhone 15 Pro - 2024-05-13 at 16 28 00

Does anyone have any idea as to why this is occurring or any context for the issue? I saw there's been some recent changes to the package in the last week or so.

**Comment by legobeat** [comment]
@Daniel-Cross needs a `yarn deduplicate`

**Comment by legobeat** [comment]
@SocketSecurity ignore npm/is-async-function@2.0.0
    @SocketSecurity ignore npm/@react-native-community/eslint-plugin@1.3.0
    @SocketSecurity ignore npm/@react-native/js-polyfills@0.72.1
    @SocketSecurity ignore npm/babel-plugin-transform-flow-enums@0.0.2

new authors ok

**Comment by legobeat** [comment]
@SocketSecurity ignore npm/@react-native-community/cli-doctor@11.4.1

shell access ok

**Comment by legobeat** [comment]
> @Daniel-Cross needs a `yarn deduplicate`

Again a `yarn dedupe` should clear the failing ci check  (:

File: patches/react-native+0.72.12.patch (hunk 3 above)
Comment (leotm) [reply patches/react-native+0.72.12.patch]: done in https://github.com/MetaMask/metamask-mobile/pull/9280/commits/40a238dfaf553e036be5c2e1450f16df2a9a6d85 and https://github.com/MetaMask/metamask-mobile/commit/6266824ea170ffc046db15d900d274a74cecf0c8

**Comment by AbuSabreen** [comment]
added14b41e252696fa4f6e5a6f63b872d00e449
https://t.me/elhadel_coin/23

**Review by AbuSabreen — COMMENTED** [review]
لا اعرف ماذا افعل

**Comment by tommasini** [comment]
@jpcloureiro quote:

"
We had silent crashed when we [upgrade react native to 0.71.6](https://github.com/MetaMask/metamask-mobile/pull/6220)
The hermes patch resolved the silent crashes.
These crashes happened when the user rejected a wallet transaction

As stated on the PR description, we cherry-picked [one hermes commit](https://github.com/facebook/hermes/commit/f2f1a2e38e86632f24176d9e607c06507f900496) that wasn't included in the react-native project at that time.
The request to include it was reject as the react-native team was encountering issues with it.
https://github.com/facebook/hermes/commit/f2f1a2e38e86632f24176d9e607c06507f900496
https://github.com/reactwg/react-native-releases/discussions/69#discussioncomment-5861539


These are all the hermes version tags that include that fix
hermes-2024-07-01-RNv0.75.0-1edbe36ce92fef2c4d427f5c4e104f2758f4b692  
hermes-2024-06-28-RNv0.74.3-7bda0c267e76d11b68a585f84cfdd65000babf85 hermes-2024-06-17-RNv0.75.0-c2c4ee7dfcf4b006268ee215b63f79e9cbeedcf4 
hermes-2024-06-03-RNv0.74.2-bb1e74fe1e95c2b5a2f4f9311152da052badc2bc hermes-2024-04-29-RNv0.73.8-644c8be78af1eae7c138fa4093fb87f0f4f8db85 
hermes-2024-04-25-RNv0.74.1-b54a3a01c531f4f5f1904cb0770033e8b7153dff hermes-2024-04-25-RNv0.73.8-e3f36ed20b0d9db08e2f3723af2a0be080778f1e 
hermes-2024-03-26-RNvmain-86d391d00be80aeccc8b29f62a254ecf6bc8a9ae hermes-2024-02-20-RNv0.74.0-999cfd9979b5f57b1269119679ab8cdf60897de9 
hermes-2024-02-20-RNv0.73.5-18f99ace4213052c5e7cdbcd39ee9766cd5df7e4 hermes-2024-01-25-RNv0.73.3-57e3665290d5e97eaea63d27cb370ce7862a10a6 
hermes-2023-11-17-RNv0.73.0-21043a3fc062be445e56a2c10ecd8be028dd9cc5 
hermes-2023-09-26-RNv0.73.0-ee2922a50fb719bdb378025d95dbd32ad93cd679

> hey @leotm a quick update on this: we tried to cherry pick it but we hit a number of issues so we'll hold off from making this Hermes change.

[Source](https://github.com/reactwg/react-native-releases/discussions/69#discussioncomment-5861539)
I don't know specifically what issues the react-native team encountered when they tried to cherry-pick.
I do remember running hermes unit test locally and failing a couple of tests.
I would leave the hermes patch as last resort.
"

**Comment by tommasini** [comment]
I Will start removing the commits that were fixing crashes and begin the steps to update to 0.73.8

**Comment by joaoloureirop** [comment]
https://github.com/MetaMask/metamask-mobile/pull/9280#issuecomment-2231126228

This convo is around Android crashes we're observing on this PR. 

@tommasini verified that the Android crash goes away after adding back the hermes patch

**Comment by tommasini** [comment]
After realizing that the hermes patch do not fix the issue!
It was identified that a loop was created with the stack of an error being added to a class the extends Error (JsonRpcError class)

Last patches fixing the crashes were reverted and it was added a patch to comment the code of the stack, added a comment to it to make sure we know why it's commented

Created a minimal reproduction for we to be able to see the issue easier:

Grabbing the stack property was creating a loop with hermes enabled
```
export default class CustomErrorr extends Error {
  constructor(code, message) {
    super(message);
    this.code = code;
  }

  serialize() {
    const serialized = {
      code: this.code,
      message: this.message,
    };
    serialized.stack = this.stack;

    return serialized;
  }

  toString() {
    console.log('ENTER CUSTOM ERROR TO STRING');
    return JSON.stringify(this.serialize());
  }
}
```

**Comment by tommasini** [comment]
@SocketSecurity ignore npm/@babel/plugin-proposal-numeric-separator@7.18.6

This is an acceptable risk, since it's part of the `metro-react-native-babel-preset` package

**Review by Cal-L — DISMISSED** [review]
Nice work, LGTM! Since Dan is OOO, we can briefly disable CLA bot to get this merged. just LMK

**Review by georgewrmarshall — DISMISSED** [review]
Approving on behalf of @MetaMask/design-system-engineers

**Review by brianacnguyen — DISMISSED** [review]

File: .e2e.env.exmaple [hunk 5]
```diff
@@ -2,4 +2,4 @@
 export MM_TEST_ACCOUNT_SRP='word1 word... word12'
 export MM_TEST_ACCOUNT_ADDRESS='0x...'
 export MM_TEST_ACCOUNT_PRIVATE_KEY=''
-export IS_TEST="true"
\ No newline at end of file
+export IS_TEST="true"
```
Comment (joaoloureirop) [inline .e2e.env.exmaple:5]: typo in this filename `exmaple`

**Review by jpuri — DISMISSED** [review]
Changes in files owned by confirmations team look great 👍

**Comment by tommasini** [comment]
@legobeat Can we sync on it? I have a fix for that conversation locally but would like to understand why it was applied on the first place

**Review by wachunei — DISMISSED** [review]
Ramp changes LGTM

**Comment by leotm** [comment]
a wild `ios/Podfile.lock` conflict appeared ^ then lgtm

**Review by andreahaku — DISMISSED** [review]
LGTM

File: android/app/build.gradle [hunk 6]
```diff
@@ -193,6 +184,14 @@ android {
...
 	  packagingOptions {
 		  exclude 'META-INF/DEPENDENCIES'
+      pickFirst 'lib/x86/libc++_shared.so'
+      pickFirst 'lib/x86_64/libc++_shared.so'
+      pickFirst 'lib/armeabi-v7a/libc++_shared.so'
+      pickFirst 'lib/arm64-v8a/libc++_shared.so'
+      pickFirst 'lib/arm64-v8a/libcrypto.so'
+      pickFirst 'lib/armeabi-v7a/libcrypto.so'
+      pickFirst 'lib/x86/libcrypto.so'
+      pickFirst 'lib/x86_64/libcrypto.so'
```
Comment (leotm) [inline android/app/build.gradle:194]: are these still needed?

Comment (tommasini) [reply android/app/build.gradle:194]: Yeah it seems that are still needed, it failes to build when removed
image

File: android/settings.gradle [hunk 7]
```diff
@@ -10,15 +10,8 @@ pluginManagement {
...
 include ':app'
-includeBuild('../node_modules/react-native-gradle-plugin')
-includeBuild('../node_modules/react-native') {
-     dependencySubstitution {
-         substitute(module("com.facebook.react:react-android")).using(project(":ReactAndroid"))
-         substitute(module("com.facebook.react:react-native")).using(project(":ReactAndroid"))
-         substitute(module("com.facebook.react:hermes-android")).using(project(":ReactAndroid:hermes-engine"))
-         substitute(module("com.facebook.react:hermes-engine")).using(project(":ReactAndroid:hermes-engine"))
-     }
-}
```
Comment (leotm) [inline android/settings.gradle:21]: nb: we can eventually remove https://github.com/MetaMask/hermes once fully no longer in use

**Review by leotm — APPROVED** [review]
lgtm ^ thanks for reviewing with me @tommasini

**Review by Cal-L — APPROVED** [review]
LGTM

**Comment by Cal-L** [comment]
Disabled CLA since Dan is OOO

**Pull request merged**